<div style="background: #3B3936">
    <h1 style="text-align:center;color:#BD2A2E">PowerCo</h1>
    <hr style="width:80%">
    <br>
    <h2 style="color:#486966;text-align:center">Info about project</h2>
    <hr style="width:30%">
    <p style="color:#B2BEBF;text-align:center">PowerCo is a major gas and electricity utility that supplies to corporate, SME (Small & Medium Enterprise), and residential customers. The power liberalization of the energy market in Europe has led to significant customer churn, especially in the SME segment. They have partnered with BCG to help diagnose the source of churning SME customers.</p>
    <h2 style="color:#486966;text-align:center">The aim of the project</h2>
        <hr style="width:30%">
    <p style="color:#B2BEBF;text-align:center">The energy market has seen a lot of change in recent years and there are more options than ever for customers to choose from. This became a big problem for PowerCo and they engaged BCG to help diagnose why their customers were swaying.</p>
    <h2 style="color:#486966;text-align:center">Datasets</h2>
    <hr style="width:30%">
    <ol> 
        <li style="font-size:20px">Historical customer data</li>
            <ul>
                <li>Customer data such as usage, subscription history, expected usage, etc</li>
            </ul>
                <table style="width:80%;margin: auto;">
                    <thead>
                        <tr>
                            <th style="border:1px solid black;padding:8px;width:30%;text-align:center;color: #BD2A2E">Column</th>
                            <th style="border:1px solid black;padding:8px;text-align:center;color: #BD2A2E">Description</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr>
                            <td style="text-align:center">id</td>
                            <td>Client company identifier</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">activity_new</td>
                            <td>Category of the company’s activity</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">channel_sales</td>
                            <td>Code of the sales channel</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">cons_12m</td>
                            <td>Electricity consumption of the past 12 months</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">cons_gas_12m</td>
                            <td>Gas consumption of the past 12 months</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">cons_last_month</td>
                            <td>Electricity consumption of the last month</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">date_activ</td>
                            <td>Date of activation of the contract</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">date_end</td>
                            <td>Registered date of the end of the contract</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">date_modif_prod</td>
                            <td>Date of the last modification of the product</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">date_renewal</td>
                            <td>Date of the next contract renewal</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">forecast_cons_12m</td>
                            <td>Forecasted electricity consumption for next 12 months</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">forecast_discount_energy</td>
                            <td>Forecasted value of current discount</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">forecast_meter_rent_12m</td>
                            <td>Forecasted bill of meter rental for the next 2 months</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">forecast_price_energy_off_peak</td>
                            <td>Forecasted energy price for 1st period (off peak)</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">forecast_price_energy_peak</td>
                            <td>Forecasted energy price for 2nd period (peak)</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">forecast_price_pow_off_peak</td>
                            <td>Forecasted power price for 1st period (off peak)</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">has_gas</td>
                            <td>Indicated if client is also a gas client</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">imp_cons</td>
                            <td>Current paid consumption</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">margin_gross_pow_ele</td>
                            <td>Gross margin on power subscription</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">margin_net_pow_ele</td>
                            <td>Net margin on power subscription</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">nb_prod_act</td>
                            <td>Number of active products and services</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">net_margin</td>
                            <td>Total net margin</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">num_years_antig</td>
                            <td>Antiquity of the client (in number of years)</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">origin_up</td>
                            <td>Code of the electricity campaign the customer first subscribed to</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">pow_max</td>
                            <td>Subscribed power</td>
                        </tr>
                        <tr>
                            <td style="text-align:center">churn</td>
                            <td>Has the client churned over the next 3 months</td>
                        </tr>
                    </tbody>
                </table>
        </li>
        <br>
        <li style="font-size:20px">Historical pricing data</li>
        <ul>
            <li>variable and fixed pricing data, etc</li>
        </ul>
        <table style="width: 80%; margin: auto;">
        <thead>
            <tr>
                <th style="border: 1px solid black; padding: 8px; color: #BD2A2E;text-align:center">Attribute</th>
                <th style="border: 1px solid black; padding: 8px; color: #BD2A2E;text-align:center">Description</th>
            </tr>
        </thead>
        <tbody style="color: #B2BEBF;">
            <tr>
                <td style="text-align: center;">id</td>
                <td>Client company identifier</td>
            </tr>
            <tr>
                <td style="text-align: center;">price_date</td>
                <td>Reference date</td>
            </tr>
            <tr>
                <td style="text-align: center;">price_off_peak_var</td>
                <td>Price of energy for the 1st period (off peak)</td>
            </tr>
            <tr>
                <td style="text-align: center;">price_peak_var</td>
                <td>Price of energy for the 2nd period (peak)</td>
            </tr>
            <tr>
                <td style="text-align: center;">price_mid_peak_var</td>
                <td>Price of energy for the 3rd period (mid peak)</td>
            </tr>
            <tr>
                <td style="text-align: center;">price_off_peak_fix</td>
                <td>Price of power for the 1st period (off peak)</td>
            </tr>
            <tr>
                <td style="text-align: center;">price_peak_fix</td>
                <td>Price of power for the 2nd period (peak)</td>
            </tr>
            <tr>
                <td style="text-align: center;">price_mid_peak_fix</td>
                <td>Price of power for the 3rd period (mid peak)</td>
            </tr>
        </tbody>
    </table>         
    </ol>
</div>


<hr style="width:95%">
<hr style="width:70%">
<hr style="width:50%">
<hr style="width:25%">
<hr style="width:10%">
<hr style="width:5%">

<div style="background:#4C4A59;width:900px;margin:auto;border-radius:8px;padding:8px">
    <h1 style="color:#DBF227;text-align:center;">Import Libraries</h1>
</div>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly_express as px
import klib

<div style="background:#4C4A59;width:900px;margin:auto;border-radius:8px;padding:8px">
    <h1 style="color:#DBF227;text-align:center;">Load data</h1>
</div>

<div style="background:#4C4A59;width:900px;margin:auto;border-radius:8px">
    <h2 style="color:#DBF227;text-align:center;">Part one  client data</h2>
</div>

In [2]:
#first we will load the client data
client=pd.read_csv('../data/external/client_data (1).csv')

In [3]:
client.shape

(14606, 26)

In [4]:
#show the 5 rows of data
client.head()

id                     channel_sales  \
0  24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   
1  d29c2c54acc38ff3c0614d0a653813dd                           MISSING   
2  764c75f661154dac3a6c254cd082ea7d  foosdfpfkusacimwkcsosbicdxkicaua   
3  bba03439a292a1e166f80264c16191cb  lmkebamcaaclubfxadlmueccxoimlema   
4  149d57cf92fc41cf94415803a877cb4b                           MISSING   

   cons_12m  cons_gas_12m  cons_last_month  date_activ    date_end  \
0         0         54946                0  2013-06-15  2016-06-15   
1      4660             0                0  2009-08-21  2016-08-30   
2       544             0                0  2010-04-16  2016-04-16   
3      1584             0                0  2010-03-30  2016-03-30   
4      4425             0              526  2010-01-13  2016-03-07   

  date_modif_prod date_renewal  forecast_cons_12m  ...  has_gas  imp_cons  \
0      2015-11-01   2015-06-23               0.00  ...        t      0.00   
1      2009-08-21   2015-08-31             189.95  ...        f      0.00   
2      2010-04-16   2015-04-17              47.96  ...        f      0.00   
3      2010-03-30   2015-03-31             240.04  ...        f      0.00   
4      2010-01-13   2015-03-09             445.75  ...        f     52.32   

   margin_gross_pow_ele  margin_net_pow_ele  nb_prod_act  net_margin  \
0                 25.44               25.44            2      678.99   
1                 16.38               16.38            1       18.89   
2                 28.60               28.60            1        6.60   
3                 30.22               30.22            1       25.46   
4                 44.91               44.91            1       47.98   

  num_years_antig                         origin_up  pow_max  churn  
0               3  lxidpiddsbxsbosboudacockeimpuepw   43.648      1  
1               6  kamkkxfxxuwbdslkwifmmcsiusiuosws   13.800      0  
2               6  kamkkxfxxuwbdslkwifmmcsiusiuosws   13.856      0  
3               6  kamkkxfxxuwbdslkwifmmcsiusiuosws   13.200      0  
4               6  kamkkxfxxuwbdslkwifmmcsiusiuosws   19.800      0  

[5 rows x 26 columns]

In [5]:
#check the name of columns
client.columns

Index(['id', 'channel_sales', 'cons_12m', 'cons_gas_12m', 'cons_last_month',
       'date_activ', 'date_end', 'date_modif_prod', 'date_renewal',
       'forecast_cons_12m', 'forecast_cons_year', 'forecast_discount_energy',
       'forecast_meter_rent_12m', 'forecast_price_energy_off_peak',
       'forecast_price_energy_peak', 'forecast_price_pow_off_peak', 'has_gas',
       'imp_cons', 'margin_gross_pow_ele', 'margin_net_pow_ele', 'nb_prod_act',
       'net_margin', 'num_years_antig', 'origin_up', 'pow_max', 'churn'],
      dtype='object')

<div style="background: #1D1E26; width: 800px; color: #D9B4B7; border-radius: 10px; margin: auto; padding: 20px;">
    <ul >
        <li>We can say that there are no problems or observations in the feature names</li>
    </ul>
</div>

In [6]:
#info about the data
client.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14606 entries, 0 to 14605
Data columns (total 26 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              14606 non-null  object 
 1   channel_sales                   14606 non-null  object 
 2   cons_12m                        14606 non-null  int64  
 3   cons_gas_12m                    14606 non-null  int64  
 4   cons_last_month                 14606 non-null  int64  
 5   date_activ                      14606 non-null  object 
 6   date_end                        14606 non-null  object 
 7   date_modif_prod                 14606 non-null  object 
 8   date_renewal                    14606 non-null  object 
 9   forecast_cons_12m               14606 non-null  float64
 10  forecast_cons_year              14606 non-null  int64  
 11  forecast_discount_energy        14606 non-null  float64
 12  forecast_meter_rent_12m         

<div style="background: #1D1E26; width: 800px; color: #D9B4B7; border-radius: 10px; margin: auto; padding: 20px;">
    <ul >
        <li>the client data contain 25 feature and 14606 rows ,the target columns is <b style="color:#F7D6D2">churn</b></li>
        <li>the feature contains 3 categorical columns and 4 data time columns and 15 numerical columns  </li>
        <li>the target columns is binary column</li>
    </ul>
</div>

In [7]:
#describe of numerical features 
client.describe().T.style.background_gradient()

<div style="background: #1D1E26; width: 800px; color: #D9B4B7; border-radius: 10px; margin: auto; padding: 20px;">
    <h3 style="text-align:center;color:#15BFAE">Summary</h3>
    <ul>
        <li><b>High Consumption:</b> The <code>cons_12m</code> (annual consumption) has a high maximum value, indicating that there are some consumers with very high energy usage. This could be due to large industrial consumers or outliers in the data.</li>
        <li><b>Gas Consumption:</b> The <code>cons_gas_12m</code> has a mean of 28092, but the 25%, 50%, and 75% quartiles are all 0. This suggests that a majority of consumers do not use gas.</li>
        <li><b>Forecasting:</b> The <code>forecast_cons_12m</code> and <code>forecast_cons_years</code> are relatively low compared to actual consumption.</li>
        <li><b>Variability:</b> The <code>std</code> is high for all consumption measures, indicating a large variability in consumption among consumers.
        <li><b>Discount Energy Forecast:</b> has a mean of 0.97, but the 25%, 50%, and 75% quartiles are all 0, this means the majority of forecasts don't include discount </li>
        <li><b> Energy Price Forecasts: </b> for the  <code>forecast_price_energy_off_peak</code> and <code>forecast_price_energy_peak</code>
        we can see that The mean values suggest that the price during peak times is generally lower than during off-peak times</li>
        <li><b> Consumption Impact</b> <code>imp_cons</code> has a high std, indicating a large variability in the impact of consumption among consumers</li>
        <li><b>Churn:</b> It has a mean of 0.097 suggesting that on average,about 9.7% of consumers stop doing business with the company within a given time period </li></li>
    </ul>
</div>
/div>
>

In [8]:
#describe of category feature
client.describe(include='object').T.style.background_gradient()

<div style="background: #1D1E26; width: 800px; color: #D9B4B7; border-radius: 10px; margin: auto; padding: 20px;">
    <h3 style="text-align: center; color: #15BFAE;">Summary</h3>
    <ul>
        <li><b>Channel Sales:</b> There are 8 unique sales channels.</li>
        <li><b>has_gas:</b> This is a binary column, with ‘f’ (presumably for ‘false’) being the most common value (11955 out of 14606).</li>
        <li><b>origin_up:</b> There are 6 unique values.</li>
    </ul>
</div>



In [9]:
#Chech the missing value 
klib.missingval_plot(client)

No missing values found in the dataset.


In [10]:
#Check the dulaplaced
client.duplicated().sum()

0

In [11]:
# check the column value
for i in client.select_dtypes('object').drop('id',axis=1):
    display(client[i].value_counts().to_frame().T)

channel_sales  foosdfpfkusacimwkcsosbicdxkicaua  MISSING  \
count                                      6754     3725   

channel_sales  lmkebamcaaclubfxadlmueccxoimlema  \
count                                      1843   

channel_sales  usilxuppasemubllopkaafesmlibmsdf  \
count                                      1375   

channel_sales  ewpakwlliwisiwduibdlfmalxowmwpci  \
count                                       893   

channel_sales  sddiedcslfslkckwlfkdpoeeailfpeds  \
count                                        11   

channel_sales  epumfxlbckeskwekxbiuasklxalciiuu  \
count                                         3   

channel_sales  fixdbufsefwooaasfcxdxadsiekoceaa  
count                                         2

date_activ  2009-08-01  2010-02-01  2009-09-01  2009-10-01  2012-02-01  \
count               95          92          76          55          48   

date_activ  2010-01-11  2010-01-18  2011-12-15  2010-06-10  2011-11-23  ...  \
count               48          47          45          42          41  ...   

date_activ  2005-03-11  2005-04-25  2005-09-07  2007-03-09  2004-04-14  \
count                1           1           1           1           1   

date_activ  2007-03-16  2005-01-28  2009-02-01  2005-02-14  2004-12-29  
count                1           1           1           1           1  

[1 rows x 1796 columns]

date_end  2016-02-01  2016-08-01  2016-09-01  2016-10-05  2016-12-31  \
count            145         125         117         115         104   

date_end  2016-10-25  2016-09-02  2016-06-10  2016-06-01  2016-03-01  ...  \
count            103          93          80          79          78  ...   

date_end  2016-10-12  2016-07-31  2016-12-25  2016-08-15  2016-10-27  \
count             10           7           6           5           4   

date_end  2017-01-29  2017-06-01  2017-01-28  2017-06-13  2017-06-11  
count              1           1           1           1           1  

[1 rows x 368 columns]

date_modif_prod  2015-11-01  2015-05-24  2015-12-17  2015-09-20  2015-12-16  \
count                   721         269         199         173         172   

date_modif_prod  2015-11-18  2015-07-11  2015-04-29  2015-08-27  2015-10-30  \
count                   164         158         155         150         149   

date_modif_prod  ...  2014-02-21  2004-01-09  2013-08-21  2010-01-02  \
count            ...           1           1           1           1   

date_modif_prod  2008-11-14  2007-01-29  2012-01-24  2010-12-28  2010-12-23  \
count                     1           1           1           1           1   

date_modif_prod  2015-05-08  
count                     1  

[1 rows x 2129 columns]

date_renewal  2015-06-23  2015-03-09  2015-02-09  2015-07-04  2015-10-11  \
count                587         451         273         265         231   

date_renewal  2015-06-14  2015-03-06  2015-06-08  2015-05-23  2015-07-09  ...  \
count                220         203         148         140         137  ...   

date_renewal  2014-06-08  2013-09-08  2014-01-13  2014-02-23  2013-09-25  \
count                  1           1           1           1           1   

date_renewal  2014-09-19  2015-01-17  2014-01-28  2013-12-22  2013-09-21  
count                  1           1           1           1           1  

[1 rows x 386 columns]

has_gas      f     t
count    11955  2651

origin_up  lxidpiddsbxsbosboudacockeimpuepw  kamkkxfxxuwbdslkwifmmcsiusiuosws  \
count                                  7097                              4294   

origin_up  ldkssxwpmemidmecebumciepifcamkci  MISSING  \
count                                  3148       64   

origin_up  usapbepcfoloekilkwsdiboslwaxobdp  ewxeelcelemmiwuafmddpobolfuxioce  
count                                     2                                 1

<div style="background: #1D1E26; width: 800px; color: #D9B4B7; border-radius: 10px; margin: auto; padding: 20px;">
    <h3 style="text-align: center; color: #15BFAE;">Summary</h3>
    <ul>
        <li>columns <code>origin_up</code> and <code>channel_sales</code> have a missing value in capital case</li>
        <li> <code> has_gas </code> has two values ``f , t `` </li>
    </ul>
</div>


In [12]:
#fixed the value of MISSING value
client['origin_up'].replace('MISSING','missing',inplace=True)
client['channel_sales'].replace('MISSING','missing',inplace=True)


<div style="background:#4C4A59;width:900px;margin:auto;border-radius:8px">
    <h2 style="color:#DBF227;text-align:center;">Part two  Price data</h2>
</div>

In [13]:
price=pd.read_csv('../data/external/price_data (1).csv')

In [14]:
#show the first 5 rwos 
price.head()

id  price_date  price_off_peak_var  \
0  038af19179925da21a25619c5a24b745  2015-01-01            0.151367   
1  038af19179925da21a25619c5a24b745  2015-02-01            0.151367   
2  038af19179925da21a25619c5a24b745  2015-03-01            0.151367   
3  038af19179925da21a25619c5a24b745  2015-04-01            0.149626   
4  038af19179925da21a25619c5a24b745  2015-05-01            0.149626   

   price_peak_var  price_mid_peak_var  price_off_peak_fix  price_peak_fix  \
0             0.0                 0.0           44.266931             0.0   
1             0.0                 0.0           44.266931             0.0   
2             0.0                 0.0           44.266931             0.0   
3             0.0                 0.0           44.266931             0.0   
4             0.0                 0.0           44.266931             0.0   

   price_mid_peak_fix  
0                 0.0  
1                 0.0  
2                 0.0  
3                 0.0  
4                 0.0

In [15]:
#the shape of data
price.shape

(193002, 8)

In [16]:
# show the columns name 
price.columns

Index(['id', 'price_date', 'price_off_peak_var', 'price_peak_var',
       'price_mid_peak_var', 'price_off_peak_fix', 'price_peak_fix',
       'price_mid_peak_fix'],
      dtype='object')

<div style="background: #1D1E26; width: 800px; color: #D9B4B7; border-radius: 10px; margin: auto; padding: 20px;">
    <ul >
        <li>in price data we  can say that there are no problems or observations in the feature names</li>
    </ul>
</div>

In [17]:
#info about the price data
price.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193002 entries, 0 to 193001
Data columns (total 8 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   id                  193002 non-null  object 
 1   price_date          193002 non-null  object 
 2   price_off_peak_var  193002 non-null  float64
 3   price_peak_var      193002 non-null  float64
 4   price_mid_peak_var  193002 non-null  float64
 5   price_off_peak_fix  193002 non-null  float64
 6   price_peak_fix      193002 non-null  float64
 7   price_mid_peak_fix  193002 non-null  float64
dtypes: float64(6), object(2)
memory usage: 11.8+ MB


<div style="background: #1D1E26; width: 800px; color: #D9B4B7; border-radius: 10px; margin: auto; padding: 20px;">
    <ul >
        <li>the price data is contain 7 features and one column is date type <code> price_date</code> and 5 columns with numerical dtype  </li>
    </ul>
</div>

In [20]:
#describe the data
price.describe().T.style.background_gradient()

<div style="background: #1D1E26; width: 800px; color: #D9B4B7; border-radius: 10px; margin: auto; padding: 20px;">
    <h3 style="text-align: center; color: #15BFAE;">Summary</h3>
    <ul>
        <li><b>Variability in Prices:</b> The standard deviation for <code>price_off_peak_var</code>, <code>price_peak_var</code>, and <code>price_mid_peak_var</code> is relatively small, indicating that the off-peak, peak, and mid-peak variable prices do not vary much.</li>
        <li>The fixed prices <code>(price_off_peak_fix, price_peak_fix, price_mid_peak_fix)</code> have larger standard deviations, suggesting more variability in these prices.</li>
        <li>The minimum, 25th percentile, and median values for <code>price_peak_var</code>, <code>price_mid_peak_var</code>, <code>price_peak_fix</code>, and <code>price_mid_peak_fix</code> are all zero. This could indicate a large number of zero values in these distributions, which might suggest that for many observations, these prices are not applicable or not charged.</li>
        <li><b>Outliers:</b> The maximum values for all variables are significantly higher than their respective 75th percentile values. This could indicate the presence of outliers in the data.</li>
        <li>The mean of <code>price_off_peak_fix</code> is significantly higher than the means of <code>price_peak_fix</code> and <code>price_mid_peak_fix</code>. This suggests that, on average, the fixed price during off-peak times is higher than during peak and mid-peak times.</li>
    </ul>
</div>
iv>


In [21]:
#check the missing value
klib.missingval_plot(price)

No missing values found in the dataset.


In [22]:
#Check the duplicate on the data
price.duplicated().sum()

0

In [24]:
#check the value of each columns 
for col in price.select_dtypes('object'):
    display(price[col].value_counts().to_frame().T)

id     038af19179925da21a25619c5a24b745  97589d95ec46ab31ef238d18a616403f  \
count                                12                                12   

id     97adbcd721fae4874f7ee287ea84200d  d74b4544363be674855cf931b6b8af25  \
count                                12                                12   

id     27276020e48d73f589788d4e37c51879  384342d542376c1bbf90d47bd31f2ba9  \
count                                12                                12   

id     edc9d08a49b565095ec33867745e5650  bbf41bc8379e2c718b8293eec8a22f5a  \
count                                12                                12   

id     bcf00c494bf6e921f508b8f401531ba8  d7d41355efdf7b0ab2ce25bc100bf9d1  \
count                                12                                12   

id     ...  9c9d8de40aee618b75f36e9b0f0586b6  \
count  ...                                 9   

id     61022f00c0541d5cf95c5b16e20db0eb  85781e805f208ddbcd9eeac081ed2615  \
count                                 9                                 9   

id     fefcf0f47956307324b63a08393127bf  223a98d3832ece78cbf279a194868b54  \
count                                 9                                 8   

id     83cf18b07114e495ae8b7fb235e45ee2  3e459d61dc831e29f8a9a9a59f95efd2  \
count                                 8                                 8   

id     c5dcd5c506e565aaabffa29bc1ec0a37  bf89f2d8c1b133a134fd93603cb4c947  \
count                                 7                                 7   

id     15b36e47cf04bf151e3f4438d12672e5  
count                                 7  

[1 rows x 16096 columns]

price_date  2015-08-01  2015-12-01  2015-07-01  2015-11-01  2015-06-01  \
count            16094       16094       16090       16087       16085   

price_date  2015-10-01  2015-02-01  2015-09-01  2015-05-01  2015-04-01  \
count            16085       16082       16082       16080       16079   

price_date  2015-03-01  2015-01-01  
count            16074       16070

<div style="background: #1D1E26; width: 800px; color: #D9B4B7; border-radius: 10px; margin: auto; padding: 20px;">
    <ul >
        <li>Some columns in the data appear as a category but are actually date type, so they must be modified  </li>
    </ul>
</div>

<div style="background:#4C4A59;width:900px;margin:auto;border-radius:8px;padding:8px">
    <h1 style="color:#DBF227;text-align:center;">Concat the two data</h1>
</div>

In [32]:
df=pd.merge(client,price,on='id')

In [33]:
df.head()

id                     channel_sales  \
0  24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   
1  24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   
2  24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   
3  24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   
4  24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   

   cons_12m  cons_gas_12m  cons_last_month  date_activ    date_end  \
0         0         54946                0  2013-06-15  2016-06-15   
1         0         54946                0  2013-06-15  2016-06-15   
2         0         54946                0  2013-06-15  2016-06-15   
3         0         54946                0  2013-06-15  2016-06-15   
4         0         54946                0  2013-06-15  2016-06-15   

  date_modif_prod date_renewal  forecast_cons_12m  ...  \
0      2015-11-01   2015-06-23                0.0  ...   
1      2015-11-01   2015-06-23                0.0  ...   
2      2015-11-01   2015-06-23                0.0  ...   
3      2015-11-01   2015-06-23                0.0  ...   
4      2015-11-01   2015-06-23                0.0  ...   

                          origin_up  pow_max  churn  price_date  \
0  lxidpiddsbxsbosboudacockeimpuepw   43.648      1  2015-01-01   
1  lxidpiddsbxsbosboudacockeimpuepw   43.648      1  2015-02-01   
2  lxidpiddsbxsbosboudacockeimpuepw   43.648      1  2015-03-01   
3  lxidpiddsbxsbosboudacockeimpuepw   43.648      1  2015-04-01   
4  lxidpiddsbxsbosboudacockeimpuepw   43.648      1  2015-05-01   

   price_off_peak_var  price_peak_var price_mid_peak_var  price_off_peak_fix  \
0            0.125976        0.103395           0.071536           40.565969   
1            0.125976        0.103395           0.071536           40.565969   
2            0.125976        0.103395           0.071536           40.565973   
3            0.125976        0.103395           0.071536           40.565973   
4            0.125976        0.103395           0.071536           40.565973   

   price_peak_fix  price_mid_peak_fix  
0       24.339581           16.226389  
1       24.339581           16.226389  
2       24.339578           16.226383  
3       24.339578           16.226383  
4       24.339578           16.226383  

[5 rows x 33 columns]

In [34]:
df.isnull().sum()

id                                0
channel_sales                     0
cons_12m                          0
cons_gas_12m                      0
cons_last_month                   0
date_activ                        0
date_end                          0
date_modif_prod                   0
date_renewal                      0
forecast_cons_12m                 0
forecast_cons_year                0
forecast_discount_energy          0
forecast_meter_rent_12m           0
forecast_price_energy_off_peak    0
forecast_price_energy_peak        0
forecast_price_pow_off_peak       0
has_gas                           0
imp_cons                          0
margin_gross_pow_ele              0
margin_net_pow_ele                0
nb_prod_act                       0
net_margin                        0
num_years_antig                   0
origin_up                         0
pow_max                           0
churn                             0
price_date                        0
price_off_peak_var          